cpp/src/plasma/lib/java/org_apache_arrow_plasma_PlasmaClientJNI.cc

```diff
@@ -0,0 +1,314 @@
+  const char* m_name = env->GetStringUTFChars(manager_socket_name, NULL);
+
+  plasma::PlasmaClient* client = new plasma::PlasmaClient();
+  ARROW_CHECK_OK(client->Connect(s_name, m_name, release_delay));
+
+  env->ReleaseStringUTFChars(store_socket_name, s_name);
+  env->ReleaseStringUTFChars(manager_socket_name, m_name);
+  return reinterpret_cast<int64_t>(client);
+}
+
+JNIEXPORT void JNICALL Java_org_apache_arrow_plasma_PlasmaClientJNI_disconnect(
+    JNIEnv* env, jclass cls, jlong conn) {
+  plasma::PlasmaClient* client = (plasma::PlasmaClient*)conn;
+
+  ARROW_CHECK_OK(client->Disconnect());
+  delete client;
+  return;
+}
+
+JNIEXPORT jobject JNICALL Java_org_apache_arrow_plasma_PlasmaClientJNI_create(
+    JNIEnv* env, jclass cls, jlong conn, jbyteArray object_id, jint size,
+    jbyteArray metadata) {
+  plasma::PlasmaClient* client = (plasma::PlasmaClient*)conn;
+  plasma::ObjectID oid;
+  jbyteArray_to_object_id(env, object_id, &oid);
+
+  // prepare metadata buffer
+  uint8_t* md = NULL;
+  jsize md_size = 0;
+  std::unique_ptr<JByteArrayGetter> md_getter;
+  if (metadata != NULL) {
+    md_size = env->GetArrayLength(metadata);
+  }
+  if (md_size > 0) {
+    md_getter.reset(new JByteArrayGetter(env, metadata, reinterpret_cast<jbyte**>(&md)));
+  }
+
+  std::shared_ptr<Buffer> data;
+  Status s = client->Create(oid, size, md, md_size, &data);
+  if (s.IsPlasmaObjectExists()) {
+    jclass Exception = env->FindClass("org/ray/spi/impl/PlasmaObjectExistsException");
+    env->ThrowNew(Exception,
+                  "An object with this ID already exists in the plasma store.");
+    return NULL;
+  }
+  if (s.IsPlasmaStoreFull()) {
+    jclass Exception = env->FindClass("org/ray/spi/impl/PlasmaOutOfMemoryException");
+    env->ThrowNew(Exception,
+                  "The plasma store ran out of memory and could not create this object.");
+    return NULL;
+  }
+  ARROW_CHECK(s.ok());
+
+  return env->NewDirectByteBuffer(data->mutable_data(), size);
+}
+
+JNIEXPORT jbyteArray JNICALL Java_org_apache_arrow_plasma_PlasmaClientJNI_hash(
+    JNIEnv* env, jclass cls, jlong conn, jbyteArray object_id) {
+  plasma::PlasmaClient* client = (plasma::PlasmaClient*)conn;
+  plasma::ObjectID oid;
+  jbyteArray_to_object_id(env, object_id, &oid);
+
+  unsigned char digest[plasma::kDigestSize];
+  bool success = client->Hash(oid, digest).ok();
+
+  if (success) {
+    jbyteArray ret = env->NewByteArray(plasma::kDigestSize);
+    env->SetByteArrayRegion(ret, 0, plasma::kDigestSize,
+                            reinterpret_cast<jbyte*>(digest));
+    return ret;
+  } else {
+    return NULL;
+  }
+}
+
+JNIEXPORT void JNICALL Java_org_apache_arrow_plasma_PlasmaClientJNI_seal(
+    JNIEnv* env, jclass cls, jlong conn, jbyteArray object_id) {
+  plasma::PlasmaClient* client = (plasma::PlasmaClient*)conn;
+  plasma::ObjectID oid;
+  jbyteArray_to_object_id(env, object_id, &oid);
+
+  ARROW_CHECK_OK(client->Seal(oid));
```

Comment:
Let's add a JIRA that we should figure out a way to propagate arrow::Status to java automatically

Reply:
ok, i will add a JIRL later